File: src_minimal/usb.c
/* usb.c — USB driver for MSM8909 minimal programmer.
 *
 * Inherits PBL's USB session — no controller reset, no re-enumeration.
 * PBL already enumerated the device and left RS=1 (D+ asserted).
 * We flush stale transfers, then use PBL's dQH table in-place.
 *
 * Critical constraint: USB DMA cannot access OCIMEM (0x0804xxxx).
 * All dTDs and data buffers must be in DDR (0x80000000+).
 * PBL's dQH table (in PBL SRAM) is used as-is — we don't swap it.
 *
 * Hardware: ChipIdea USB 2.0 HS controller at 0x078D9000.
 */

#include "usb.h"
#include "msm8909.h"

#define REG32(addr)  (*(volatile unsigned int *)(addr))
#define readl(addr)  REG32(addr)
#define writel(v, addr) do { REG32(addr) = (v); } while(0)

/*========================================================================
 * Cache maintenance — required for USB DMA coherency
 *========================================================================*/
static void cache_flush_inv(void *addr, unsigned int len)
{
    unsigned int line;
    unsigned int start = (unsigned int)addr & ~31u;
    unsigned int end = (unsigned int)addr + len;
    for (line = start; line < end; line += 32)
        __asm__ volatile("mcr p15, 0, %0, c7, c14, 1" :: "r"(line));
    __asm__ volatile("dsb" ::: "memory");
}

static void cache_inv(void *addr, unsigned int len)
{
    unsigned int line;
    unsigned int start = (unsigned int)addr & ~31u;
    unsigned int end = (unsigned int)addr + len;
    for (line = start; line < end; line += 32)
        __asm__ volatile("mcr p15, 0, %0, c7, c6, 1" :: "r"(line));
    __asm__ volatile("dsb" ::: "memory");
}

/*========================================================================
 * DDR-resident buffers — USB DMA can only access DDR, not OCIMEM.
 * These go in .ddr_bss section (mapped to DDR region in linker script).
 *========================================================================*/

/* dTDs for bulk endpoints — one per direction */
static struct ept_queue_item dtd_bulk_out __attribute__((aligned(32), section(".ddr_bss")));
static struct ept_queue_item dtd_bulk_in  __attribute__((aligned(32), section(".ddr_bss")));

/* Staging buffers for DMA — callers may pass OCIMEM pointers */
#define USB_BUF_SIZE  4096
static unsigned char usb_rx_buf[USB_BUF_SIZE] __attribute__((aligned(32), section(".ddr_bss")));
static unsigned char usb_tx_buf[USB_BUF_SIZE] __attribute__((aligned(32), section(".ddr_bss")));

/*========================================================================
 * State
 *========================================================================*/
static int usb_online;

/* Pointers to PBL's dQH entries for EP1 bulk endpoints */
static struct ept_queue_head *dqh_ep1_out;  /* index 2: EP1 OUT (host→device) */
static struct ept_queue_head *dqh_ep1_in;   /* index 3: EP1 IN  (device→host) */

/*========================================================================
 * Internal: queue a transfer on a PBL dQH entry
 *========================================================================*/
static void queue_transfer(struct ept_queue_head *dqh,
                           struct ept_queue_item *dtd,
                           void *buf, int len)
{
    unsigned int buf_addr = (unsigned int)buf;

    /* Set up dTD */
    dtd->next  = TERMINATE;
    dtd->info  = INFO_BYTES(len) | INFO_IOC | INFO_ACTIVE;
    dtd->page0 = buf_addr;
    dtd->page1 = (buf_addr & 0xFFFFF000u) + 0x1000;
    dtd->page2 = (buf_addr & 0xFFFFF000u) + 0x2000;
    dtd->page3 = (buf_addr & 0xFFFFF000u) + 0x3000;
    dtd->page4 = (buf_addr & 0xFFFFF000u) + 0x4000;
    cache_flush_inv(dtd, sizeof(*dtd));

    /* Link dTD into PBL's dQH */
    dqh->next = (unsigned int)dtd;
    dqh->info = 0;  /* clear overlay status */
    cache_flush_inv(dqh, sizeof(*dqh));
}

/*========================================================================
 * usb_init — Inherit PBL's USB session
 *
 * PBL has fully enumerated the device for Sahara and left:
 *   - RS=1 (controller running, D+ asserted)
 *   - Active transfers primed (ENDPTSTAT non-zero)
 *   - dQH table in PBL SRAM (not OCIMEM)
 *   - Device address assigned, EP1 configured
 *
 * We just flush stale transfers and prepare for bulk I/O.
 *========================================================================*/
void usb_init(void)
{
    unsigned int pbl_eplist;

    usb_online = 0;

    /* 1. Read PBL's ENDPOINTLISTADDR to find dQH table */
    pbl_eplist = readl(USB_ENDPOINTLISTADDR);

    /* dQH entries: index 2 = EP1 OUT, index 3 = EP1 IN (64 bytes each) */
    dqh_ep1_out = (struct ept_queue_head *)(pbl_eplist + 2 * 64);
    dqh_ep1_in  = (struct ept_queue_head *)(pbl_eplist + 3 * 64);

    /* 2. Disable USB interrupts (PBL may have them enabled) */
    writel(0, USB_USBINTR);

    /* 3. Flush all endpoints (clear PBL's pending Sahara transfers) */
    writel(0xFFFFFFFF, USB_ENDPTFLUSH);
    while (readl(USB_ENDPTFLUSH));

    /* 4. Clear pending status bits */
    writel(readl(USB_USBSTS), USB_USBSTS);
    writel(readl(USB_ENDPTCOMPLETE), USB_ENDPTCOMPLETE);

    /* Online immediately — PBL already enumerated, host handle is live */
    usb_online = 1;
}

/*========================================================================
 * usb_read — Blocking bulk OUT read on EP1
 *
 * DMA goes to DDR staging buffer, then memcpy to caller's buffer.
 * Returns number of bytes received, or -1 on error.
 *========================================================================*/
int usb_read(void *buf, int maxlen)
{
    unsigned int bit = EPT_RX(1);
    int xfer_len, remaining, actual, i;

    if (!usb_online)
        return -1;
    if (maxlen <= 0)
        return 0;

    /* Cap to staging buffer size */
    xfer_len = maxlen;
    if (xfer_len > USB_BUF_SIZE)
        xfer_len = USB_BUF_SIZE;

    /* Invalidate rx staging buffer */
    cache_flush_inv(usb_rx_buf, xfer_len);

    /* Queue bulk OUT transfer on PBL's dQH[2] */
    queue_transfer(dqh_ep1_out, &dtd_bulk_out, usb_rx_buf, xfer_len);
    writel(bit, USB_ENDPTPRIME);

    /* Poll for completion */
    for (;;) {
        unsigned int sts = readl(USB_USBSTS);

        if (sts & STS_URI) {
            /* USB bus reset — session lost */
            writel(STS_URI, USB_USBSTS);
            usb_online = 0;
            return -1;
        }

        if (sts & (STS_UI | STS_UEI)) {
            writel(sts & (STS_UI | STS_UEI), USB_USBSTS);

            if (readl(USB_ENDPTCOMPLETE) & bit) {
                writel(bit, USB_ENDPTCOMPLETE);
                break;
            }
        }
    }

    /* Read actual transfer size from dTD */
    cache_inv(&dtd_bulk_out, sizeof(dtd_bulk_out));
    remaining = (dtd_bulk_out.info >> 16) & 0x7FFF;
    actual = xfer_len - remaining;

    /* Copy from DDR staging buffer to caller's buffer */
    cache_inv(usb_rx_buf, actual);
    {
        unsigned char *dst = (unsigned char *)buf;
        for (i = 0; i < actual; i++)
            dst[i] = usb_rx_buf[i];
    }

    return actual;
}

/*========================================================================
 * usb_write — Blocking bulk IN write on EP1
 *
 * Copies caller's data to DDR staging buffer, then DMA from there.
 * Returns number of bytes sent, or -1 on error.
 *========================================================================*/
int usb_write(const void *buf, int len)
{
    unsigned int bit = EPT_TX(1);
    int xfer_len, remaining, i;

    if (!usb_online)
        return -1;
    if (len <= 0)
        return 0;

    /* Cap to staging buffer size */
    xfer_len = len;
    if (xfer_len > USB_BUF_SIZE)
        xfer_len = USB_BUF_SIZE;

    /* Copy caller's data to DDR staging buffer */
    {
        const unsigned char *src = (const unsigned char *)buf;
        for (i = 0; i < xfer_len; i++)
            usb_tx_buf[i] = src[i];
    }
    cache_flush_inv(usb_tx_buf, xfer_len);

    /* Queue bulk IN transfer on PBL's dQH[3] */
    queue_transfer(dqh_ep1_in, &dtd_bulk_in, usb_tx_buf, xfer_len);
    writel(bit, USB_ENDPTPRIME);

    /* Poll for completion */
    for (;;) {
        unsigned int sts = readl(USB_USBSTS);

        if (sts & STS_URI) {
            writel(STS_URI, USB_USBSTS);
            usb_online = 0;
            return -1;
        }

        if (sts & (STS_UI | STS_UEI)) {
            writel(sts & (STS_UI | STS_UEI), USB_USBSTS);

            if (readl(USB_ENDPTCOMPLETE) & bit) {
                writel(bit, USB_ENDPTCOMPLETE);
                break;
            }
        }
    }

    cache_inv(&dtd_bulk_in, sizeof(dtd_bulk_in));
    remaining = (dtd_bulk_in.info >> 16) & 0x7FFF;
    return xfer_len - remaining;
}

File: src_rust/src/usb.rs
/// USB bulk I/O — inherits PBL's USB session.
///
/// Port of usb.c. All DMA buffers in DDR via link_section.
/// USB DMA cannot access OCIMEM (0x0804xxxx).
use crate::mmio::Reg32;
use crate::platform::{MSM_USB_BASE, dsb};

// Register addresses
const USB_USBSTS: u32 = MSM_USB_BASE + 0x0144;
const USB_USBINTR: u32 = MSM_USB_BASE + 0x0148;
const USB_ENDPOINTLISTADDR: u32 = MSM_USB_BASE + 0x0158;
const USB_ENDPTPRIME: u32 = MSM_USB_BASE + 0x01B0;
const USB_ENDPTFLUSH: u32 = MSM_USB_BASE + 0x01B4;
const USB_ENDPTSTAT: u32 = MSM_USB_BASE + 0x01B8;
const USB_ENDPTCOMPLETE: u32 = MSM_USB_BASE + 0x01BC;

// Endpoint bits
const EPT_RX_1: u32 = 1 << 1;
const EPT_TX_1: u32 = 1 << 17;

// Status bits
const STS_URI: u32 = 1 << 6;

// dTD constants
const TERMINATE: u32 = 1;
const INFO_IOC: u32 = 1 << 15;
const INFO_ACTIVE: u32 = 1 << 7;

const USB_BUF_SIZE: usize = 4096;

/// dTD (device Transfer Descriptor) — 32 bytes, 32-byte aligned.
#[repr(C, align(32))]
struct EptQueueItem {
    next: u32,
    info: u32,
    page0: u32,
    page1: u32,
    page2: u32,
    page3: u32,
    page4: u32,
    _reserved: u32,
}

/// DDR-resident buffers — USB DMA can only access DDR.
#[link_section = ".ddr_bss"]
static mut DTD_BULK_OUT: EptQueueItem = EptQueueItem {
    next: 0, info: 0, page0: 0, page1: 0, page2: 0, page3: 0, page4: 0, _reserved: 0,
};
#[link_section = ".ddr_bss"]
static mut DTD_BULK_IN: EptQueueItem = EptQueueItem {
    next: 0, info: 0, page0: 0, page1: 0, page2: 0, page3: 0, page4: 0, _reserved: 0,
};

#[link_section = ".ddr_bss"]
static mut USB_RX_BUF: [u8; USB_BUF_SIZE] = [0u8; USB_BUF_SIZE];
#[link_section = ".ddr_bss"]
static mut USB_TX_BUF: [u8; USB_BUF_SIZE] = [0u8; USB_BUF_SIZE];

static mut USB_ONLINE: bool = false;
static mut DQH_EP1_OUT: u32 = 0; // address of dQH for EP1 OUT
static mut DQH_EP1_IN: u32 = 0;  // address of dQH for EP1 IN

fn readl(addr: u32) -> u32 { Reg32::new(addr).read() }
fn writel(val: u32, addr: u32) { Reg32::new(addr).write(val); }

/// Prime an endpoint: set up dTD, link into dQH, write ENDPTPRIME.
unsafe fn prime_ep(dqh_addr: u32, dtd: *mut EptQueueItem, buf_addr: u32, len: u32, prime_bit: u32) {
    // Wait for previous prime
    while readl(USB_ENDPTPRIME) & prime_bit != 0 {}

    // Flush if active
    if readl(USB_ENDPTSTAT) & prime_bit != 0 {
        writel(prime_bit, USB_ENDPTFLUSH);
        while readl(USB_ENDPTFLUSH) & prime_bit != 0 {}
    }

    // Clear stale completion
    if readl(USB_ENDPTCOMPLETE) & prime_bit != 0 {
        writel(prime_bit, USB_ENDPTCOMPLETE);
    }

    // Set up dTD
    (*dtd).next = TERMINATE;
    (*dtd).info = (len << 16) | INFO_IOC | INFO_ACTIVE;
    (*dtd).page0 = buf_addr;
    (*dtd).page1 = (buf_addr & 0xFFFF_F000) + 0x1000;
    (*dtd).page2 = (buf_addr & 0xFFFF_F000) + 0x2000;
    (*dtd).page3 = (buf_addr & 0xFFFF_F000) + 0x3000;
    (*dtd).page4 = (buf_addr & 0xFFFF_F000) + 0x4000;
    dsb();

    // Link into dQH: current=0, next=dtd, info=0
    let dqh = dqh_addr as *mut u32;
    // dQH layout: [0]=config, [1]=current, [2]=next, [3]=info
    core::ptr::write_volatile(dqh.add(1), 0); // current
    core::ptr::write_volatile(dqh.add(2), dtd as u32); // next
    core::ptr::write_volatile(dqh.add(3), 0); // info (clear overlay)
    dsb();

    // Prime
    writel(prime_bit, USB_ENDPTPRIME);
}

/// Poll until transfer completes or bus reset. Returns true on success.
unsafe fn poll_complete(bit: u32) -> bool {
    loop {
        if readl(USB_USBSTS) & STS_URI != 0 {
            writel(STS_URI, USB_USBSTS);
            USB_ONLINE = false;
            return false;
        }
        if readl(USB_ENDPTCOMPLETE) & bit != 0 {
            writel(bit, USB_ENDPTCOMPLETE);
            return true;
        }
    }
}

/// Initialize USB — inherit PBL's session.
pub fn init() {
    unsafe {
        USB_ONLINE = false;

        let pbl_eplist = readl(USB_ENDPOINTLISTADDR);
        DQH_EP1_OUT = pbl_eplist + 2 * 64;
        DQH_EP1_IN = pbl_eplist + 3 * 64;

        // Disable interrupts
        writel(0, USB_USBINTR);

        // Flush all endpoints
        writel(0xFFFF_FFFF, USB_ENDPTFLUSH);
        while readl(USB_ENDPTFLUSH) != 0 {}

        // Clear status
        writel(readl(USB_USBSTS), USB_USBSTS);
        writel(readl(USB_ENDPTCOMPLETE), USB_ENDPTCOMPLETE);

        USB_ONLINE = true;
    }
}

/// Blocking bulk OUT read on EP1.
pub fn read(buf: &mut [u8]) -> i32 {
    unsafe {
        if !USB_ONLINE { return -1; }
        if buf.is_empty() { return 0; }

        let xfer_len = buf.len().min(USB_BUF_SIZE) as u32;

        prime_ep(DQH_EP1_OUT, &raw mut DTD_BULK_OUT,
                 USB_RX_BUF.as_ptr() as u32, xfer_len, EPT_RX_1);

        if !poll_complete(EPT_RX_1) { return -1; }

        dsb();
        let remaining = (DTD_BULK_OUT.info >> 16) & 0x7FFF;
        let actual = (xfer_len - remaining) as usize;

        for i in 0..actual {
            buf[i] = USB_RX_BUF[i];
        }
        actual as i32
    }
}

/// Blocking bulk IN write on EP1.
pub fn write(buf: &[u8]) -> i32 {
    unsafe {
        if !USB_ONLINE { return -1; }
        if buf.is_empty() { return 0; }

        let xfer_len = buf.len().min(USB_BUF_SIZE) as u32;

        for i in 0..xfer_len as usize {
            USB_TX_BUF[i] = buf[i];
        }
        dsb();

        prime_ep(DQH_EP1_IN, &raw mut DTD_BULK_IN,
                 USB_TX_BUF.as_ptr() as u32, xfer_len, EPT_TX_1);

        if !poll_complete(EPT_TX_1) { return -1; }

        dsb();
        let remaining = (DTD_BULK_IN.info >> 16) & 0x7FFF;
        (xfer_len - remaining) as i32
    }
}

/// Write a string slice over USB.
pub fn write_str(s: &str) -> i32 {
    write(s.as_bytes())
}
